File: backend/app/macro/registry.py
# -*- coding: utf-8 -*-
"""
Macro Registry - registri makro.

Mendefinisikan himpunan nama makro bawaan beserta meta-informasinya.
"""

from dataclasses import dataclass


@dataclass
class MacroMeta:
    """
    Meta-informasi makro.

    Attributes:
        name: Nama makro.
        configurable: Apakah dapat dikonfigurasi di panel samping.
        description: Deskripsi.
        handler_class: Nama kelas handler (dimuat secara lazy).
    """

    name: str
    configurable: bool
    description: str
    handler_class: str


MACRO_REGISTRY: dict[str, MacroMeta] = {
    "getmem": MacroMeta(
        name="getmem",
        configurable=False,
        description="Mengambil isi memori bab",
        handler_class="GetMemHandler",
    ),
    "getlist": MacroMeta(
        name="getlist",
        configurable=False,
        description="Mengambil daftar isi 50 bab terbaru",
        handler_class="GetListHandler",
    ),
    "getworld": MacroMeta(
        name="getworld",
        configurable=False,
        description="Mengambil entri buku dunia proyek saat ini yang permanen dan cocok kata kunci",
        handler_class="GetWorldHandler",
    ),
    "if": MacroMeta(
        name="if",
        configurable=True,
        description="Awal blok render kondisional (berbasis variabel bool)",
        handler_class="IfHandler",
    ),
    "endif": MacroMeta(
        name="endif",
        configurable=True,
        description="Akhir blok render kondisional",
        handler_class="EndIfHandler",
    ),
}


def get_macro_names() -> set[str]:
    """Mengambil semua nama makro bawaan."""
    return set(MACRO_REGISTRY.keys())


def is_valid_macro(name: str) -> bool:
    """Memeriksa apakah nama makro valid."""
    return name in MACRO_REGISTRY


def is_configurable(name: str) -> bool:
    """Memeriksa apakah makro dapat dikonfigurasi."""
    meta = MACRO_REGISTRY.get(name)
    return meta.configurable if meta else False

File: frontend/src/lib/macro/registry.ts
/**
 * Macro Registry - 宏注册表
 *
 * 定义预定义的宏名集合和元信息。
 */

import type { MacroMeta } from "./types";

export const MACRO_REGISTRY: Record<string, MacroMeta> = {
  getmem: {
    name: "getmem",
    configurable: false,
    description: "获取章节记忆内容",
  },
  getlist: {
    name: "getlist",
    configurable: false,
    description: "获取最新 50 个章节的目录列表",
  },
  getworld: {
    name: "getworld",
    configurable: false,
    description: "获取当前项目世界书常驻和关键词命中条目",
  },
  if: {
    name: "if",
    configurable: true,
    description: "条件渲染块开始（基于bool变量）",
  },
  endif: {
    name: "endif",
    configurable: true,
    description: "条件渲染块结束",
  },
};

/** 获取所有预定义宏名 */
export function getMacroNames(): Set<string> {
  return new Set(Object.keys(MACRO_REGISTRY));
}

/** 检查宏名是否有效 */
export function isValidMacro(name: string): boolean {
  return name in MACRO_REGISTRY;
}

/** 检查宏是否可配置 */
export function isConfigurable(name: string): boolean {
  const meta = MACRO_REGISTRY[name];
  return meta?.configurable ?? false;
}

/** 获取宏元信息 */
export function getMacroMeta(name: string): MacroMeta | undefined {
  return MACRO_REGISTRY[name];
}
